File: device/software/tools/meta.py
# -*- coding: utf-8 -*-
import os
import sys
import struct

#print sys.argv[1:]
categories = ["Other", "Agriculture", "Health", "Education", "Stories", "Business", "Governance", "Music", "Diary"]

if len(sys.argv) == 1:
  print "usage: python " + sys.argv[0] + " .a18 file names\n" 
  print "  for example: python " + sys.argv[0] + " f1.a18 f2.a18 ...\n"

for fn in sys.argv[1:]:
  print ' '
  f = os.open(fn, os.O_RDONLY)
  alen = os.read(f, 4)
  value = struct.unpack('<i4', alen)
  pos = os.lseek(f, value[0] + 4, os.SEEK_SET)
  print fn + "has " + str(value[0]) + " audio bytes"
#  print pos
  metaver = alen = os.read(f, 4)
  value = struct.unpack('<i4', metaver)
  print "  metedata version=" + str(value[0])
  numf = os.read(f, 4)
  value = struct.unpack('<i4', numf)
  numfields = value[0]
  print "  num fields=" + str(numfields)
  
  myfieldIdx = 0
  while myfieldIdx < numfields:
    fid = os.read(f, 2)
    value = struct.unpack('<h2', fid)
    field_id = value[0]
    print "\n    field id=" + str(field_id)
    
    flen = os.read(f, 4)
    field_len = struct.unpack('<i4', flen)[0]
    print "      field length=" + str(field_len)
    
    nfv = os.read(f, 1)
    num_f_values = struct.unpack('<b', nfv)
    num_field_values = num_f_values[0]
    print "      num field values=" + str(num_field_values)
    
    myvalueIdx = 0
    while myvalueIdx < num_field_values:
      fln = os.read(f, 2)
      value = struct.unpack('<h2', fln)
      fvalue_len = value[0]
      print "      field value length=" + str(fvalue_len)
      fvalue = os.read(f, fvalue_len)
      print    "      field value[" + str(myvalueIdx) + ']=' + str(fvalue)
      myvalueIdx = myvalueIdx + 1
      
  
    myfieldIdx = myfieldIdx + 1
    
  os.close(f)
    

File: device/software/tools/meta.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>

char *categories[] = {"Other", "Agriculture", "Health", "Education", "Stories", "Business", "Governance", "Music", "Diary" };

int 
main(int argc, char*argv[])
{
	long wrk, numfields, fieldlen;
	int ret;
	short fid;
	unsigned char nfv;
	unsigned char buf[128];
	int i, j, k, n;

	if(argc == 1) {
		printf("usage: %s .a18 file names\n", argv[0]);
		printf("  for example: %s f1.a18 f2.a18 ...\n", argv[0]);
	}

	for(n = 1; n <= (argc - 1); n++) {

	int fd = open(argv[n], 0);
	if(fd <= 0) {
		fprintf(stderr, "can't open %s\n",argv[n]);
	}
	ret = read(fd, &wrk, 4);
	printf("%s has %ld audio bytes\n",argv[n], wrk);
	wrk = lseek(fd, wrk + 4, SEEK_SET);
	ret = read(fd, &wrk, 4);
	printf("meta data version=%ld\n", wrk);
	ret = read(fd, &numfields, 4);
	printf("num fields=%ld\n", numfields);

	for(i=0; i<(int)numfields; i++) {
		ret = read(fd, &fid, 2);
		printf("\n  field id=%d\n",fid);
		ret = read(fd, &fieldlen, 4);
		printf("    filed length=%d\n", fieldlen);
		ret = read(fd, &nfv, 1);
		printf("    num field values=%d\n", nfv);
		for(j=0; j<nfv; j++) {
			short fl;
			ret = read(fd, &fl, 2);
			printf("    field value length[%d]=%d\n",j,fl);
			ret = read(fd, buf, fl);
			buf[fl] = 0;
			printf("    field value[%d]=",j);
/*			for(k=0; k<fl; k++) {
				printf("0x%.2x ", buf[k]);
			}
			printf("\n");
*/
			printf("'%s'",buf);
			if(fid == 0) { // categories
				unsigned int m = buf[0] - '0';
				if((m >= 0 && m < 9)) {
					printf(" (%s) ", categories[m]);
				}
			}
			printf("\n");
		}
	}
	close(fd);
	printf("\n\n");
	}
}
